### DISCUSSION TITLE
DNS C2 Spec

### DISCUSSION THREAD

**Discussion by its-a-feature** [its-a-feature/Mythic#418] [Ideas]
On the roadmap for Mythic is to create a DNS C2 profile, so I'd like to start a discussion as to the features, requirements, and specifications people would like to see in it before we start the development process

**Comment by v1stra** [comment]
I think it would be great to be able to specify different record types to be used, (TXT, AAAA, A), etc

**Reply by its-a-feature** [reply]
Absolutely! It would be great to make sure they match too. So if you're requesting an A record, you get back a standard response. Having the request type being dynamically configurable would be really helpful too so that you could use something like A records normally, but for large output or file transfers switch to TXT

**Comment by nollium** [comment]
A lot of dns servers (e.g 8.8.8.8) randomize the case of dns queries to purposefully break dns tunnels using base64.

An option to use base32 (case insensitive) instead is a must have

**Reply by its-a-feature** [reply]
Oh, I hadn't heard of that before! That's super good to know! We could also use something like the netbios encoding used by Cobalt Strike (https://unit42.paloaltonetworks.com/cobalt-strike-metadata-encoding-decoding/) to constrain the character set and always lowercase it so that the sensitivity doesn't matter

**Reply by v1stra** [reply]
DNS-0x20 encoding randomly capitalizes the characters https://astrolavos.gatech.edu/articles/increased_dns_resistance.pdf

**Comment by nikallass** [comment]
Does existing DNS C2 profile working?
https://github.com/MythicC2Profiles/dns

**Reply by its-a-feature** [reply]
That hasn't been updated, no. It's still on my roadmap to do, other things just came up that needed to happen first

**Comment by piolug93** [comment]
Possiblity use multiple domain for receive one message. That for evade detection based on volume requests for one domain.

**Reply by pr0b3r7** [reply]
To be clear, are you referring to rotational hosts in the same way it is being discussed [here](https://github.com/its-a-feature/Mythic/discussions/423#discussion-7329047)? If so, +1 on my end.

**Reply by piolug93** [reply]
Yes this is exactly what I had in mind.

**Comment by piolug93** [comment]
Minimalize sended data e.g. payloadUUID/callbackUUID is long and consume plce for date in small dns message storage.

**Comment by its-a-feature** [comment]
The dns profile is now released in a beta state along with support for it in the Poseidon agent. If you pull the latest of those two you can test it out and see if it meets all the things we've discussed here so far. The only thing I'm tracking that _isn't_ included yet is DoH, but I want to make sure to get the standard DNS piece how we want first, then tackle that part.
